### ISSUE TITLE
[ML-v0.2] Développement d'un modèle LSTM universel multi-actifs pour prédictions robustes

### ISSUE THREAD

**Issue by Y0nyx** [Y0nyx/Agentic-Trader#16]
### 🎯 Objectif
Développer un modèle LSTM universel capable de prédire les prix de multiples actifs financiers (stocks, ETFs, crypto) en s'entraînant sur un dataset diversifié pour créer un modèle plus robuste et versatile.

### 🧠 Hypothèse
L'entraînement sur plusieurs actifs différents pourrait créer un modèle plus versatile pour plusieurs raisons :

#### ✅ **Avantages potentiels :**
1. **Généralisation accrue** : Le modèle apprend des patterns communs plutôt que spécifiques à un actif
2. **Robustesse aux régimes de marché** : Exposition à différents cycles et volatilités
3. **Réduction de l'overfitting** : Plus de données diverses = moins de sur-apprentissage
4. **Transfer learning** : Patterns appris sur un actif peuvent s'appliquer à d'autres

#### ⚠️ **Défis techniques :**
1. **Normalisation cross-assets** : Gérer des échelles de prix très différentes
2. **Saisonnalité différente** : Crypto 24/7 vs marchés traditionnels
3. **Corrélations variables** : Crypto vs actions vs ETFs
4. **Feature engineering** : Indicateurs pertinents pour tous types d'actifs

#### 1. **Dataset multi-actifs enrichi**
- [ ] Minimum 10 actifs diversifiés :
  ```python
  assets = [
      # Crypto
      "BTC-USD", "ETH-USD", "ADA-USD",
      # Tech stocks
      "AAPL", "GOOGL", "MSFT", "NVDA",
      # ETFs
      "SPY", "QQQ", "VTI"
  ]
  ```

#### 2. **Preprocessing universel**
- [ ] Normalisation robuste cross-assets :
  ```python
  # Features normalisées :
  - Returns (log returns pour comparabilité)
  - Volatilité rolling (standardisée)
  - Volume relatif (% du volume moyen)
  - RSI, MACD (déjà 0-100 ou centrés)
  ```

#### 3. **Architecture LSTM adaptée**
- [ ] Modèle capable de gérer la diversité :
  ```python
  # Architecture suggérée :
  - Feature embedding layer (pour encoder le type d'actif)
  - LSTM layers avec attention mechanism
  - Dropout élevé (0.3-0.4) pour éviter overfitting
  - Output head adaptatif selon le type d'actif
  ```

#### 4. **Validation comparative**
- [ ] Comparer les performances :
  - Modèle universel vs modèles spécialisés par actif
  - Métriques : RMSE, précision directionnelle, Sharpe ratio
  - Test sur actifs non vus pendant l'entraînement

#### 5. **Transfer learning et fine-tuning**
- [ ] Possibilité d'adapter le modèle universel :
  ```python
  # Pipeline suggéré :
  1. Pré-entraînement sur dataset multi-actifs
  2. Fine-tuning sur actif spécifique
  3. Évaluation des gains de performance
  ```

#### Tests A/B requis
```python
def test_universal_vs_specialized():
    """Compare modèle universel vs spécialisés"""
    for asset in test_assets:
        universal_score = universal_model.evaluate(asset_data[asset])
        specialized_score = specialized_models[asset].evaluate(asset_data[asset])
        
        assert universal_score['directional_accuracy'] >= specialized_score['directional_accuracy'] * 0.9
        # Accepter 10% de perte max pour la versatilité

def test_cross_asset_generalization():
    """Test sur actifs non vus pendant l'entraînement"""
    unseen_assets = ["TSLA", "DOGE-USD"]
    for asset in unseen_assets:
        score = universal_model.evaluate(asset_data[asset])
        assert score['directional_accuracy'] > 0.52  # Mieux que le hasard
```

### 📊 Features engineering multi-actifs

```python
def create_universal_features(df: pd.DataFrame, asset_type: str) -> pd.DataFrame:
    """
    Crée des features universelles normalisées.
    
    Args:
        df: OHLCV data
        asset_type: 'crypto', 'stock', 'etf'
    """
    features = pd.DataFrame()
    
    # Features universelles
    features['log_returns'] = np.log(df['close'] / df['close'].shift(1))
    features['volatility_20d'] = features['log_returns'].rolling(20).std()
    features['rsi'] = calculate_rsi(df['close'])
    features['volume_ratio'] = df['volume'] / df['volume'].rolling(20).mean()
    
    # Asset type encoding
    features['asset_type_crypto'] = 1 if asset_type == 'crypto' else 0
    features['asset_type_stock'] = 1 if asset_type == 'stock' else 0
    features['asset_type_etf'] = 1 if asset_type == 'etf' else 0
    
    return features
```

### 🧪 Exemple d'utilisation
```python
from ml_models.universal_lstm import UniversalLSTMPredictor
from ml_models.multi_asset_preprocessing import prepare_multi_asset_data

# 1. Préparation du dataset multi-actifs
assets_data = prepare_multi_asset_data([
    "BTC-USD", "ETH-USD", "AAPL", "GOOGL", "SPY", "QQQ"
])

# 2. Entraînement du modèle universel
universal_model = UniversalLSTMPredictor(
    sequence_length=60,
    n_features=10,  # Features universelles + asset type
    use_attention=True
)

history = universal_model.train(assets_data, epochs=200)

# 3. Évaluation cross-asset
for asset in ["TSLA", "DOGE-USD"]:  # Actifs non vus
    predictions = universal_model.predict(asset_data[asset])
    metrics = evaluate_predictions(predictions, actual_prices)
    print(f"{asset} - Précision directionnelle: {metrics['directional_accuracy']:.2%}")

# 4. Fine-tuning sur actif spécifique
btc_specialized = universal_model.fine_tune("BTC-USD", btc_data, epochs=50)
```

### 🎯 Métriques de succès
- **Précision directionnelle cross-asset** : > 55%
- **Performance vs modèles spécialisés** : -10% max (trade-off versatilité)
- **Généralisation sur actifs non vus** : > 52%
- **Réduction de l'overfitting** : Validation loss stable

### 🔄 Stratégie expérimentale
1. **Phase 1** : Baseline - modèles spécialisés par actif
2. **Phase 2** : Modèle universel simple
3. **Phase 3** : Modèle universel avec attention + asset encoding
4. **Phase 4** : Fine-tuning et optimisation

### 📈 Recherche et références
- Études sur universal financial models
- Transfer learning en finance
- Multi-task learning pour prédictions temporelles

---

**Excellente intuition ! L'approche multi-actifs pourrait effectivement créer un modèle plus robuste en apprenant des patterns communs aux marchés financiers plutôt qu'en se sur-spécialisant sur un seul actif.**
